feat: TrackPropagation backwards when needed

wdconinc opened the pull request:
### Briefly, what does this PR introduce?
This PR causes TrackPropagation to use backwards propagation when the track tip is on a surface that is larger than the target surface.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue: https://indico.bnl.gov/event/28479/contributions/108745/attachments/62657/107619/06_05_2025_AngularResoutions_Tracking_WG.pdf, p27)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No.

### Does this PR change default behavior?
No.

wdconinc committed:
feat: TrackPropagation backwards when needed

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc committed:
fix: isValid only since 36.1.0

wdconinc commented:
No diffs for the same surfaces in the config.

wdconinc committed:
feat: propagate to DIRC surface

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc commented:
After adding the DIRC effective surface with ID 90, the hits do indeed show up there. Still all forward propagated until we include the BIC hits.

wdconinc committed:
fix: iwyu

wdconinc committed:
fix: do it the way clang-tidy likes it

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bschmookler committed:
Merge branch 'main' into track-propagation-backwards-when-needed

bschmookler committed:
Removed Acts v39 references

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bschmookler commented:
I updated the branch to not use the references to Acts v.39 (which were already removed from the main branch).

I tested the projections to the barrel ECal, and I can now see projections to both surfaces. This follows up on the presentation here: https://indico.bnl.gov/event/32582/contributions/123516/attachments/69639/119480/Observations%20for%20track%20projections%20to%20calorimeters.pdf

```events->Draw("hypot(_CalorimeterTrackProjections_points.position.x,_CalorimeterTrackProjections_points.position.y):_CalorimeterTrackProjections_points.position.z","_CalorimeterTrackProjections_points.system==101","*")```

<br/><br/>

With default track projection (main branch):

<img width="697" height="477" alt="default" src="https://github.com/user-attachments/assets/47cfdfb8-e8b4-4b3d-bcca-98d3ba9663bc" />

<br/><br/>

With new (backwards-compatible) track projection:

<img width="699" height="476" alt="new" src="https://github.com/user-attachments/assets/d594f860-0409-4279-b3ec-d9fb183090a1" />

bschmookler commented:
With this branch, both surfaces are seen for each track projection:

```
root [1] events->Scan("_CalorimeterTrackProjections_points.surface","_CalorimeterTrackProjections_points.system==101")
***********************************
*    Row   * Instance * _Calorime *
***********************************
*        0 *        3 *         1 *
*        0 *        4 *         2 *
*        2 *        1 *         1 *
*        2 *        2 *         2 *
*        3 *        1 *         1 *
*        3 *        2 *         2 *
*        8 *        1 *         1 *
*        8 *        2 *         2 *
*        9 *        1 *         1 *
*        9 *        2 *         2 *
*       11 *        0 *         1 *
*       11 *        1 *         2 *
*       12 *        1 *         1 *
*       12 *        2 *         2 *
*       13 *        0 *         1 *
*       13 *        1 *         2 *
*       14 *        1 *         1 *
*       14 *        2 *         2 *
*       15 *        1 *         1 *
*       15 *        2 *         2 *
```

epic-capybara committed:
feat: TrackPropagation backwards when needed (fix: iwyu) (#2651)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/25328925508.
Please merge this PR into the branch
`track-propagation-backwards-when-needed`
to resolve failures in PR #1921.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler reviewed (approved):
Thanks for this fix @bschmookler and @wdconinc ! I'm very happy to see the BIC projections!

From my end, this is ready to go! Copilot's suggestion on reusing `initPosition` and `initDirection` is reasonable, but isn't critical. And I'm seeing the [eicrecon-DIS (clang++, 18x275, 0, craterlake_18x275, TSAN, 4)](https://github.com/eic/EICrecon/actions/runs/25331042408/job/74266438913?pr=1921#logs) check fail due to the readout not being defined for the pfRICH, but I've seen that in a couple other PRs so I don't think it's due to this one!

veprbl reviewed (approved):
LGTM

veprbl committed:
Merge branch 'main' into track-propagation-backwards-when-needed

bschmookler committed:
Merge branch 'main' into track-propagation-backwards-when-needed
